dart/lib/src/protocol/access_aware_map.dart

```diff
@@ -0,0 +1,50 @@
+import 'dart:collection';
+
+import 'package:meta/meta.dart';
+
+@internal
+class AccessAwareMap<String, V> extends MapBase<String, V> {
+  AccessAwareMap(this._map);
+
+  final Map<String, V> _map;
+  final Set<String> _accessedKeys = {};
+
+  Set<String> get accessedKeys => _accessedKeys;
+
+  @override
+  V? operator [](Object? key) {
+    if (key is String) {
+      _accessedKeys.add(key);
+    }
+    return _map[key];
+  }
+
+  @override
+  void operator []=(String key, V value) {
+    _map[key] = value;
+  }
+
+  @override
+  void clear() {
+    _map.clear();
+    _accessedKeys.clear();
+  }
+
+  @override
+  Iterable<String> get keys => _map.keys;
+
+  @override
+  V? remove(Object? key) {
+    return _map.remove(key);
+  }
+
+  Map<String, dynamic>? notAccessed() {
+    Map<String, dynamic> unknown = _map.keys
+        .where((key) => !accessedKeys.contains(key))
+        .fold<Map<String, dynamic>>({}, (map, key) {
+      map[key] = _map[key];
+      return map;
+    });
+    return unknown.isNotEmpty ? unknown : null;
```

Comment:
do you have a rough idea of the impact of this? since we always instantiate a new map and iterate through the keys

I would imagine it's small since the jsons are generally not large

Reply:
This can be improved for the majority of cases by the trivial check before anything else:
```suggestion
    if (accessedKeys.length == _map.length) {
        return null;
    }
    Map<String, dynamic> unknown = _map.keys
        .where((key) => !accessedKeys.contains(key))
        .fold<Map<String, dynamic>>({}, (map, key) {
      map[key] = _map[key];
      return map;
    });
    return unknown.isNotEmpty ? unknown : null;
```